File: apps/frontend/src/shared/api/cursorPage.ts
import { z } from 'zod';

// 커서 기반 페이지. items 한 페이지 + 다음 커서. hasNext=false 면 nextCursor=null.
// totalCount: 현재 조건에 매칭되는 전체 수. 첫 페이지에서만 채워지고 이후엔 null인 경우가 많다
// (매 페이지 COUNT 를 피하려는 최적화). 도메인별 item 스키마를 받아 페이지 스키마를 만든다.
export const cursorPageSchema = <T extends z.ZodTypeAny>(item: T) =>
  z.object({
    items: z.array(item),
    nextCursor: z.string().nullable(),
    hasNext: z.boolean(),
    totalCount: z.number().int().nonnegative().nullish(),
  });

File: apps/backend/src/main/java/com/airdnd/room/dto/CursorPage.java
package com.airdnd.room.dto;

import java.util.List;

// 커서 기반 한 페이지. hasNext=false 면 nextCursor 는 null 이고 더 가져올 항목이 없다.
// totalCount 는 현재 조건(영역+필터)에 매칭되는 전체 수로, 첫 페이지에서만 채워지고 이후 페이지는 null
// ("이 지역에 N곳 — 더 좁혀보세요" 안내용). 매번 COUNT 하지 않으려는 최적화다.
public record CursorPage<T>(
        List<T> items,
        String nextCursor,
        boolean hasNext,
        Long totalCount
) {
}
